wpilibc/src/test/native/cpp/TracerTest.cpp

```diff
@@ -0,0 +1,210 @@
+  frc::Tracer::SetThreadName("TracerTest");
+
+  frc::sim::PauseTiming();
+
+  frc::Tracer::StartTrace("Test1");
+  frc::Tracer::TraceFunc("Test2", []() { frc::sim::StepTiming(400_ms); });
+  frc::sim::StepTiming(100_ms);
+  frc::Tracer::EndTrace();
+
+  auto test1Entry = nt::NetworkTableInstance::GetDefault().GetEntry(
+      "/Tracer/TracerTest/Test1");
+  auto test2Entry = nt::NetworkTableInstance::GetDefault().GetEntry(
+      "/Tracer/TracerTest/Test1/Test2");
+
+  EXPECT_NEAR(test1Entry.GetDouble(0.0), 500.0, 1.0);
+  EXPECT_NEAR(test2Entry.GetDouble(0.0), 400.0, 1.0);
+
+  frc::sim::ResumeTiming();
+  frc::Tracer::ResetForTest();
+}
+
+TEST(TracerTest, ScopedTrace) {
+  frc::Tracer::ResetForTest();
+  frc::Tracer::SetThreadName("ScopedTracerTest");
+
+  frc::sim::PauseTiming();
+
+  frc::Tracer::StartTrace("Test3");
+  {
+    auto st = frc::Tracer::StartScopedTrace("Test4");
+    frc::sim::StepTiming(400_ms);
+  }
+  frc::sim::StepTiming(100_ms);
+  frc::Tracer::EndTrace();
+
+  auto test3Entry = nt::NetworkTableInstance::GetDefault().GetEntry(
+      "/Tracer/ScopedTracerTest/Test3");
+  auto test4Entry = nt::NetworkTableInstance::GetDefault().GetEntry(
+      "/Tracer/ScopedTracerTest/Test3/Test4");
+
+  EXPECT_TRUE(test3Entry.Exists());
+  EXPECT_TRUE(test4Entry.Exists());
+  EXPECT_NEAR(test3Entry.GetDouble(0.0), 500.0, 1.0);
+  EXPECT_NEAR(test4Entry.GetDouble(0.0), 400.0, 1.0);
+
+  frc::sim::ResumeTiming();
+  frc::Tracer::ResetForTest();
+}
+
+TEST(TracerTest, Threaded) {
+  frc::Tracer::ResetForTest();
+  frc::Tracer::SetThreadName("TracerTest");
+
+  frc::sim::PauseTiming();
+
+  // run a trace in the main thread, assert that the tracer ran
+  {
+    frc::Tracer::StartTrace("ThreadTest1");
+    frc::sim::StepTiming(100_ms);
+    frc::Tracer::EndTrace();
+
+    auto test1Entry = nt::NetworkTableInstance::GetDefault().GetEntry(
+        "/Tracer/TracerTest/ThreadTest1");
+
+    EXPECT_TRUE(test1Entry.Exists());
+    EXPECT_NEAR(test1Entry.GetDouble(0.0), 100.0, 1.0);
+  }
+
+  // run a trace in a new thread, assert that the tracer ran
+  // and that the output position and value are correct
+  {
+    std::thread thread([]() {
+      frc::Tracer::SetThreadName("TestThread");
+      frc::Tracer::StartTrace("ThreadTest1");
+      frc::sim::StepTiming(400_ms);
+      frc::Tracer::EndTrace();
+
+      auto test1Entry = nt::NetworkTableInstance::GetDefault().GetEntry(
+        "/Tracer/TracerTest/ThreadTest1");
+
+      EXPECT_TRUE(test1Entry.Exists());
+      EXPECT_NEAR(test1Entry.GetDouble(0.0), 100.0, 1.0);
+
+      auto test2Entry = nt::NetworkTableInstance::GetDefault().GetEntry(
+        "/Tracer/TestThread/ThreadTest1");
+
+      EXPECT_TRUE(test2Entry.Exists());
+      EXPECT_NEAR(test2Entry.GetDouble(0.0), 400.0, 1.0);
+    });
+    thread.join();
+
+    auto test2Entry = nt::NetworkTableInstance::GetDefault().GetEntry(
+      "/Tracer/TestThread/ThreadTest1");
+
+    EXPECT_TRUE(test2Entry.Exists());
```

Comment:
@PeterJohnson These two assertions fail yet the same assertions inside the thread succeed. I don't believe it's a timing issue, I tried adding a sleep after the join. I have debug printing to make sure all the tracer states are using the same NT instance.

Reply:
It appears to be something about the publishers being dropped/destroyed causing the values to no longer be accessible in NT. Looking at the code I can't find where the drop code occurs but even setting the topics to `retained` doesn't fix this issue. Also somehow doing a `std::swap` on the thread_local `TracerState` drops all the publishers in the original state?